tools/iree-run-mlir-main.cc

```diff
@@ -209,396 +84,327 @@ std::string InferDefaultDeviceFromBackend(const std::string& backend) {
+  std::vector<char*> runtime_args_;
+  std::vector<char*> compiler_args_;
+};
 
 }  // namespace
 
-extern "C" int main(int argc_llvm, char** argv_llvm) {
+extern "C" int main(int argc, char** argv) {
   IREE_TRACE_SCOPE0("iree-run-mlir");
 
-  mlir::DialectRegistry registry;
-  mlir::iree_compiler::registerAllDialects(registry);
-  mlir::iree_compiler::registerHALTargetBackends();
-  mlir::iree_compiler::registerVMTargets();
-  mlir::registerBuiltinDialectTranslation(registry);
-  mlir::registerLLVMDialectTranslation(registry);
-  // Make sure command line options are registered.
-  // Flag options structs (must resolve prior to CLI parsing).
-  (void)mlir::iree_compiler::BindingOptions::FromFlags::get();
-  (void)mlir::iree_compiler::InputDialectOptions::FromFlags::get();
-  (void)mlir::iree_compiler::HighLevelOptimizationOptions::FromFlags::get();
-  (void)mlir::iree_compiler::SchedulingOptions::FromFlags::get();
-  (void)mlir::iree_compiler::IREE::HAL::TargetOptions::FromFlags::get();
-  (void)mlir::iree_compiler::IREE::VM::TargetOptions::FromFlags::get();
-  (void)mlir::iree_compiler::IREE::VM::BytecodeTargetOptions::FromFlags::get();
-
-  // Register MLIRContext command-line options like
-  // -mlir-print-op-on-diagnostic.
-  mlir::registerMLIRContextCLOptions();
-  // Register assembly printer command-line options like
-  // -mlir-print-op-generic.
-  mlir::registerAsmPrinterCLOptions();
-  // Register pass manager command-line options like -mlir-print-ir-*.
-  mlir::registerPassManagerCLOptions();
-
-  // On Windows InitLLVM re-queries the command line from Windows directly and
-  // totally messes up the array.
-  llvm::setBugReportMsg(
-      "Please report issues to https://github.com/openxla/iree/issues and "
-      "include the crash backtrace.\n");
-  llvm::InitLLVM init_llvm(argc_llvm, argv_llvm);
-  llvm::cl::ParseCommandLineOptions(argc_llvm, argv_llvm);
-
-  // Consume all options after the positional filename and pass them to the IREE
-  // flag parser.
-  std::vector<char*> argv_iree = {argv_llvm[0]};
-  for (auto& run_arg : run_args_flag) {
-    if (run_arg == "--") continue;
-    argv_iree.push_back(const_cast<char*>(run_arg.c_str()));
+  // Initialize the compiler once on startup before using any other APIs.
+  ireeCompilerGlobalInitialize();
+
+  // Parse full argument list and split into compiler/runtime flag sets.
+  ArgParser arg_parser;
+  if (!arg_parser.Parse(argc, argv)) {
+    ireeCompilerGlobalShutdown();
+    return 1;
   }
-  int argc_iree = static_cast<int>(argv_iree.size());
-  char** argv_iree_ptr = argv_iree.data();
-  iree_flags_parse_checked(IREE_FLAGS_PARSE_MODE_DEFAULT, &argc_iree,
-                           &argv_iree_ptr);
-
-  auto status = RunFile(input_file_flag, registry);
-  if (!status.ok()) {
-    fprintf(stderr, "ERROR running file (%s):\n%s\n", input_file_flag.c_str(),
-            status.ToString().c_str());
+
+  // Pass along compiler flags.
```

Comment:
Maybe add: "Since this is a command line tool, we initialize the global compiler command line environment. In-process/library uses would usually not do this and would set session specific arguments as needed."